### PR TITLE
headless process approach

### PR DESCRIPTION
**Pull request by hkmoon**
Hi all,

Currently, I am working on headless process for SPIM pipeline. And, this pull request should be WIP(Work In Progress). However, I would like to let you how things going on for this approach.

This pull request will contain Unit tests for IJ1Macro, API and CLI.
IJ1Macro is for Fiji Users, API is for advanced developers who want to use this APIs in any other cluster environment e.g. massage passing interface, hadoop and so on. Finally, CLI is for the users who are interested in only specific tasks by command line input. Furthermore, this approach will be benefit for conversion to IJ2Plugin as well as KNIME.

In order to handle these multiple interfaces, I would like to use ```Parameters``` static inner class in each plugin for separating GUI and process logic. Actually, this ```Parameters``` pattern is used BigDataServer by @tpietzsch and me, which is very useful such a case.

So far, I have completed ```LightSheetZ1``` for this purpose, please, feel free to comment on it. The next one is ```Resave_HDF5```(However, this class has ```Parameters```). It will not take too long. Then, ```Interest_Point_Detection```, ```Interest_Point_Registration``` and ```Image_Fusion``` are following. Still, I am collecting some use-cases to understand how to split/merge for cluster jobs.

The below checked items are taken into account for headless mode. Please, check or add more if I miss some tasks that should be included for headless support.

### 1. Define_Multi_View_Dataset
  - [x] LightSheetZ1
  - [ ] StackListLOCI
  - [x] StackListImageJ
  - [ ] MicroManger

### 2. Resave
  - [x] Resave_HDF5
  - [ ] Resave_TIFF

### 3. Interest_Point_Detection
  - [x] DifferenceOfMean
  - [x] DifferenceOfGaussian

### 4. Interest_Point_Registration
  - [x] GeometricHashing
  - [x] RGLDM
  - [x] ItereativeClosestPoint

#### 5.1 Compute on:
  - [x] CPU
  - [x] GPU

#### 5.2 Fusion Algorithm:
  - [x] EfficientBayesianBased
  - [x] WeightedAverageFusion with FUSEDATA
  - [ ] WeightedAverageFusion with INDEPENDENT

#### 5.3 BoundingBox algorithm:
  - [ ] ManualBoundingBox
  - [ ] BigDataViewerBoundingBox
  - [ ] AutomaticReorientation
  - [ ] AutomaticBoundingBox

#### 5.4 ImageExport option:
  - [ ] DisplayImage
  - [x] Save3dTIFF
  - [ ] ExportSpimData2TIFF
  - [ ] ExportSpimData2HDF5
  - [ ] AppendSpimData2

Please, let me know if there are additional tasks to support headless mode urgently other than those I mentioned above.

Thanks a lot!

### CODE REVIEW

**Comment by StephanPreibisch** [comment]
Hi @hkmoon,

great progress ... I am having a look now. By the way, for the fusion you should only allow the PreDefinedBoundingBox. This makes is way easier.

Cheers,
Stephan

File: pom.xml [hunk 1]
```diff
@@ -151,8 +155,21 @@
 		<dependency>
 			<groupId>java3d</groupId>
 			<artifactId>vecmath</artifactId>
-			<scope>provided</scope>
```
Comment (StephanPreibisch) [inline pom.xml]: Why did you change that?

File: src/main/java/spim/fiji/datasetmanager/LightSheetZ1.java [hunk 2]
```diff
@@ -1,11 +1,13 @@
 package spim.fiji.datasetmanager;
 
 import fiji.util.gui.GenericDialogPlus;
+import ij.IJ;
 import ij.gui.GenericDialog;
 
-import java.awt.Font;
+import java.awt.*;
```
Comment (StephanPreibisch) [inline src/main/java/spim/fiji/datasetmanager/LightSheetZ1.java]: Please do not do these import * things ... I like to know what exactly is required.

**Comment by StephanPreibisch** [comment]
Hi @hkmoon,

first of all thanks for your efforts so far, it is really great that you care for this project and are motivated. And I am really sorry about this, but after looking at it in in more detail I am a bit confused, I assumed after our emails that you would implement it for the ImageJ Macro Language (not actually changing my code), creating an interface structure that allows my source code to be extended without changing your cluster interface. Maybe this is also my fault for not discussing it better with you what we actually want, but I am really super busy because I leave New York May 31st for Berlin having to do a lot of experiments.

I feel that the amount of changes you are now trying to do to my source code will not easily be doable and is not well extendable as far as I can see. I do not intend to de-motivate you, but if you actually want to significantly change my source code we need to discuss a strategy first and agree on how to do that so there are as little as possible conflicts during development.

I also feel that the way you are doing it right now seems too fast and not really thought through, which is not really your fault because it is really hard making the right abstraction decisions without fully understanding the entire code and its parameters. For example duplicating the source code for the defaultProcess methods is not good idea, and also the parameters for the defaultProcess methods for example in LightSheetZ1.java or in Interest_Point_Registration and Image_Fusion are not a good choice in my opinion.

Again, don’t get me wrong, it is really great that you invest time in this and are motivated about it. And I think we will manage to make it work together. However, we need to coordinate things and I see two ways here how to do that right. We could

a) develop a plan together on how to do things right and then start, which will take time
or
b) you can start to develop an external infrastructure using the ImageJ Macro language as basis, once this works we can replace the macro language calls by making direct calls into the code and adapting it piece by piece on the way, starting with the modules that are most stable. This makes you more independent, obviously. And it will way faster yield results that work. And ultimately, this will include a)

So let’s maybe try to talk about this on Skype and make a decision?
Thanks again for your efforts,
Stephan

**Comment by emmenlau** [comment]
Dear Stephan and all,
I'm involved in a project with the goal to automate SPIM fusion without any manual interaction or GUI on a cluster environment. It seems hkmoon (and possibly others?) are going in the same direction. Should we join forces? What is the right forum to summarize the status and document next steps?

**Comment by StephanPreibisch** [comment]
This is great! I suggest that we meet for a small hackathon that I would organize at my new institute (https://www.mdc-berlin.de/13800178/en/bimsb) in Berlin for a few days where we decide how to do go forward. The integration into KNIME is closely related to it, so it makes sense to do this in one meeting I think. We should aim for June/July to do that, depending on availability of everyone.

What do you think about that @tpietzsch @hkmoon @emmenlau @dietzc

**Comment by hkmoon** [comment]
Hi @StephanPreibisch,

> first of all thanks for your efforts so far, it is really great that you care for this project and are motivated. And I am really sorry about this, but after looking at it in in more detail I am a bit confused, I assumed after our emails that you would implement it for the ImageJ Macro Language (not actually changing my code), creating an interface structure that allows my source code to be extended without changing your cluster interface. Maybe this is also my fault for not discussing it better with you what we actually want, but I am really super busy because I leave New York May 31st for Berlin having to do a lot of experiments.

No problem, Stephan! My intention of the pull request is to find out the best way to encompass all the extensions which might be coming later. So, please, do not worry even though this pull request might contain a full of crap codes and misunderstandings. I would like to use this PR as a message center for our headless approach conversations. Anyway, we would like to adopt ImageJ2 Plugin model as far as I know. That is why I have decided to change your codes (sorry that I did not let you know previously, that is actually my fault) for scalability.
> 
> I feel that the amount of changes you are now trying to do to my source code will not easily be doable and is not well extendable as far as I can see. I do not intend to de-motivate you, but if you actually want to significantly change my source code we need to discuss a strategy first and agree on how to do that so there are as little as possible conflicts during development.

Please, never mind all the changes except ```spim.fiji.datasetmanager``` package where ```Parameters``` class is introduced in ```LightSheetZ1``` for your future plugin development strategy to separate logic and UI.
> 
> I also feel that the way you are doing it right now seems too fast and not really thought through, which is not really your fault because it is really hard making the right abstraction decisions without fully understanding the entire code and its parameters. For example duplicating the source code for the defaultProcess methods is not good idea, and also the parameters for the defaultProcess methods for example in LightSheetZ1.java or in Interest_Point_Registration and Image_Fusion are not a good choice in my opinion.

I am so sorry again because I should have explained it before. ```defaultProcess()``` is made (only for me) to understand your tasks and have them run with only default parameters without asking any dialogue. That is just proof-of-concept codes that should be removed. I would like to decide to how to handle headless operations in an optimal way with you. Please, do not put too much time on my codes now. I just wanted to know which way is the best way to extend your pipeline easily for the future usage. So, it would be great to share other people's ideas here and reduce the confusions by communication. I do not think ImageJ1 Macro interface is a proper idea for your long-term plan. However, this approach will guarantee the backward compatibility.
> 
> Again, don’t get me wrong, it is really great that you invest time in this and are motivated about it. And I think we will manage to make it work together. However, we need to coordinate things and I see two ways here how to do that right. We could

Thank you so much, Stephan! I really appreciate your comments and explaining your concerns.
> 
> a) develop a plan together on how to do things right and then start, which will take time
> or
> b) you can start to develop an external infrastructure using the ImageJ Macro language as basis, once this works we can replace the macro language calls by making direct calls into the code and adapting it piece by piece on the way, starting with the modules that are most stable. This makes you more independent, obviously. And it will way faster yield results that work. And ultimately, this will include a)

I would take b) option because many people are running your pipeline in the cluster already. Then, let's do b) first including a). And, simultaneously, let's not forget refactoring the previous codes too!
Currently, @schmiedc has bash macros. I can implement interface classes to encapsulate IJMacros by referencing his codes. In this case, your codes will not be touched at all. ;)
> 
> So let’s maybe try to talk about this on Skype and make a decision?

How about discussing this topic here and mail just before 31st May 2015? I do not want to interrupt you that much until you will come to Berlin. 

Thanks a lot again, Stephan!
HongKee

**Comment by dietzc** [comment]
@StephanPreibisch The idea with the hackathon sound great. As we already discussed, for me the week from 29th of June to 03rd of July would be perfect.

@hkmoon @StephanPreibisch @emmenlau concerning the goal of this hackathon or the headless execution of the plugin in general: Without being to specific, I see several options how we can make the headless execution of parts of the plugin and therefore a KNIME integration happen. However, any of these options, involves touching your code @StephanPreibisch as we have to completely decouple parts of the plugins (for example DoG, Fusion Algorithm etc) from the UI _and_ from the data-format and persistence mechanism (including storing/loading of data). However, the open question which remains from my point of view, is how we implement these algorithms in an extendible and clean way:

* Option A is, that the Algorithms of SPIM_Registration stay independent from whatever framework and simply offers access to these algorithms. Pretty similar to imglib2-algorithms (static access to algorithms for example, ExecutionContext can be set from outside etc.). Like this, we could easily write wrappers for ImageJ-Ops and ImageJ2. 
* Option B: Algorithms in the SPIM_Registration make directly use of the modular and extensible ImageJ-Ops framework, which would solve the 'extensible' part and additionally nicely integrates with other existing ops (shared context, ThreadService, Status Service, GPU detection etc). ImageJ2 wrappers can easily be added.
* Option C: We let SPIM_Registration completely rely on ImageJ-Ops and/or ImageJ2 (automatic UI generation for IJ2, KNIME and potentially others in the future).

For a nice KNIME integration any of these options would be O.K. as we can always wrap everything into ImageJ-Ops and ImageJ2 without losing any performance. @ctrueden maybe you can elaborate a bit more on the advantages / disadvantages of these options?

However, in the end @StephanPreibisch you have to decide what ever you want! I'm looking forward to a potential hackathon in Berlin!

Christian

**Comment by emmenlau** [comment]
Great idea with the hackaton, count me in! I'll need to get some funding but I'm confident it will work out.
Goals from my side: The final "solution" for me must be fully self-dependent and run with zero user interaction, zero popup dialogues etc, on a cluster/grid/cloud. The input and output must be exclusively limited to files and/or stdout/stderr messages. Any required parameterization should ideally be performed before the start of the processing. Only in case it's strictly not possible, I could live with "late" parameterization during processing, however this would require a full abort-parameterize-resume step on the cluster, so its relatively "expensive" for me.
I have already an open source cluster workflow manager that integrates with cluster queuing systems, and does reliable staging of inputs and outputs, chaining of processing steps, execution of the processes, etc. so this part can be considered "solved" from my side.
Does that sound reasonable? Do you foresee problems?

**Comment by dietzc** [comment]
@emmenlau this sounds exactly like what we want to achieve with a KNIME integration. From a cluster execution point of view: Which cluster engine are you planning to use (Torque, SGE, SLURM, LSF?). Maybe UNICORE is something you may want to take a look at. From a name point of view, at the moment a SGE binding is supported, which means once integrated into KNIME, SPIM Reg would run headless on SGE (Sun Grid Engine).  However, we are currently evaluating UNICORE and other systems to potentially support these, too.

**Comment by schmiedc** [comment]
Hi guys,

Just to let you know. @psteinb,@StephanJanosch and me are implementing the plugin now using snakemake (https://bitbucket.org/johanneskoester/snakemake/wiki/Home) and have a collaboration with UNICORE to get a GUI solution for that as well. 

Cheers,
Christopher

**Comment by emmenlau** [comment]
Sounds very interesting! My workflow manager (BeeWM, http://screeningbee.org/) currently supports SGE/UGE (via DRMAA 1) and LSF. Support for others can be added relatively easily. I see that there is a DRMAA-based target system interface framework for UNICORE, so that might just work as well. But in any case, I'm very open to other cluster solutions, and a strong future goal for me would be cloud support.

**Comment by dietzc** [comment]
@schmiedc what does 'implementing' mean? Reimplementing the code from @StephanPreibisch such that headless execution is better supported or reusing what is already there and running some scripted version of it somewhere?

**Comment by schmiedc** [comment]
@dietzc we are reusing what is already there and run beanshell scripts.

**Comment by StephanPreibisch** [comment]
Just to be more precise, the major goals for the hackathon would be:
- decouple GUI and processing in SPIM_Registration, so that parameters can be provided or queried by any kind of input mechanism (ImageJ1, ImageJ2, KNIME, ...), so similar to option A) of what @dietzc suggested. 
- make the basic processing algorithms used (convolution etc) more abstract so that calling either GPU, ImgLib2, ImageJ-OPS is easily exchangeable.
- solve the XML problem (how to combine multiple changes concurrently done to the same XML)

Would the dates 29th of June to 03rd of July work? I have to check, but I should be able to provide accommodation ...

**Comment by psteinb** [comment]
+1 (at least) for this! the separation that is discussed here would make 
integrating
https://github.com/psteinb/libmultiviewnative or similar for a better 
GPU exploitation much easier.

**Comment by hkmoon** [comment]
I think the most time consuming task is decoupling. Would it be a good idea to collect some nice use-cases/patterns/examples how to deal with decoupling GUI and Processing here before the hackathon? Then, we can easily grab the common ways to do in the hackathon and it will be more productive.

> decouple GUI and processing in SPIM_Registration, so that parameters can be provided or queried by any kind of input mechanism (ImageJ1, ImageJ2, KNIME, ...), so similar to option A) of what @dietzc suggested.
Actually, my goad is to make ```SPIM_Registration``` run with ```-Djava.awt.headless=true``` like what @emmenlau mentioned previously. I also agree with option A) of @dietzc. Then, we can achieve higher scalability later (e.g. OpenMPI implementation of SPIM_Registration). @dietzc Where can I find ```ExecutionContext ``` examples?

> make the basic processing algorithms used (convolution etc) more abstract so that calling either GPU, ImgLib2, ImageJ-OPS is easily exchangeable.
How about making the basic processing algorithms thread-safe as well, in order to solve the XML problem at the same time. Then, we can easily build the test environments for the below topic and might bring the good solution easily.

> solve the XML problem (how to combine multiple changes concurrently done to the same XML)

@StephanPreibisch: There is one more thing (if possible) to do is to make unit-test per task and integration-test classes per cases thoroughly. How about trying TDD(Test Driven Development) in this time?

> Would the dates 29th of June to 03rd of July work? I have to check, but I should be able to provide accommodation ...
I can join you from the night of 30th of June because our service facility has a schedule until that day.

**Comment by StephanJanosch** [comment]
@emmenlau please have a look there https://github.com/StephanJanosch/singleStepPipeline/blob/master/snakefiles/Snakefile

As @schmiedc mentioned, we are working here on automated way of using @StephanPreibisch 's plugin and currently i am quite happy with the results.

**Comment by dietzc** [comment]
@hkmoon a nice example for the usage of ExecutionContext and general access to an algorithm can be found at https://github.com/imglib/imglib2-algorithm/blob/master/src/main/java/net/imglib2/algorithm/gauss3/Gauss3.java

**Comment by hkmoon** [comment]
@dietzc Thanks! I would say this is a really good example for me. ExecutionContext can be used for the algorithms based on *CPU or GPU*.

**Comment by StephanPreibisch** [comment]
@schmiedc @StephanJanosch @psteinb If you you are interested in attending the hackathon as well just let me know, might be useful, you are of course welcome ... it seems the dates will be June 29th until July 2nd/3rd. I just don't want to to be too big as this usually does not foster productivity.

**Comment by psteinb** [comment]
Go ahead with the hackathon and concentrate on the API/GUI split, I can't come for family reasons. I'd profit big time though for integrating native (GPU) libs later on.

I also wanted to share this new repo of ours that was mentioned somewhere above:
https://github.com/mpicbg-scicomp/snakemake-workflows
We have @schmiedc workflow in there for now, there are maybe more to come. 

[Snakemake](https://bitbucket.org/johanneskoester/snakemake/wiki/Home) seems to be a really versatile tool, not only restricted to HPC environments.

Good luck moving into Berlin! Congrats!

**Comment by schmiedc** [comment]
@StephanPreibisch I would like to join.

**Comment by hkmoon** [comment]
Hi all,

I am closing this PR because that approach does not seem to work. It would be good to try another approach not to or less touch the core codes. In such a case, I would ask your opinions. Also, ImageJ2 parameters are macro recordable. That should be considered too. In order to use this, the version of pom-bigdataviewer needs to be 2.1.4 containing @ctrueden's changes (https://github.com/imagej/imagej-legacy/commit/02e9e46296a9fd44962b3cd7ccda12d53183e704).

@StephanPreibisch How are you doing in Berlin? I am looking forward to seeing you there soon.

**Comment by StephanJanosch** [comment]
@StephanPreibisch seems I will come too to work on tests (of detection and registration) and other things. Any details about the hackathon itself yet?

**Comment by StephanPreibisch** [comment]
I just arrived in Berlin and starting to organize, the week of June 30th will work out, details to come in the next two days ...

**Comment by StephanJanosch** [comment]
Cool! Looking forward to this!

**Comment by schmiedc** [comment]
@StephanPreibisch @tpietzsch I would like to mention on thing: the plugin largely does not abide to the thread limitation one can set with Fiji. I tested this and only define dataset, xml merge and timelapse registration are respecting this. This is a huge problem for me and also for Stephan. The pipeline is entirely unstable with this bug. Is there any way you can solve that on your side?

**Comment by StephanJanosch** [comment]
For me it's a minor issue, because I can use the GPU for detection and deconvolution, so CPU loads are smaller. As i resave into zip, this is also not a problem, because this is slow but uses really only one core.

@schmiedc Let's have a look at libsysconfcpu. This might fix all java related core usage issue and solves the problem for every java based app on the cluster.

**Comment by schmiedc** [comment]
@StephanJanosch good

**Comment by StephanPreibisch** [comment]
I fixed that quite some time ago and tested it on my machine ... can you give an example where it does not work?

**Comment by schmiedc** [comment]
Resaving into hdf5, detection and/or registration, content based fusion and deconvolution are not sticking to the settings. 

This is how I set it (maybe that is wrong): 

import ij.Prefs;  
Prefs.setThreads(7); 

This works as I said for defining the dataset, xml merge and time lapse registration. 
I do detection and registration in one step so it is possible that registration sticks to it. 

For me it is still kind of workable since I have less than a hundred time points but there are people with a lot of time points that will saturate the cluster. Then its game over and any automation will fail since the jobs become unpredictable.

**Comment by tpietzsch** [comment]
@schmiedc @StephanPreibisch My bad.... I think this is a problem in the BigDataViewer HDF5 export code. I don't pay attention to thread limits there. I'll work on fixing that.

**Comment by schmiedc** [comment]
@tpietzsch if would be so great if that is taken care of :) :)

**Pull request closed without merging**